tools/tsp-client/src/index.ts

```diff
@@ -0,0 +1,104 @@
+import * as path from "node:path";
+
+import { createPackageJson, installDependencies } from "./npm.js";
+import { createTempDirectory, removeDirectory, tryReadTspLocation, writeFileTree } from "./fs.js";
+import { Logger, printBanner, enableDebug, printVersion } from "./log.js";
+import { doesFileExist, downloadTsp, isValidUrl, rewriteGitHubUrl } from "./network.js";
+import { compileTsp } from "./typespec.js";
+import { getEmitterPackage } from "./languageSettings.js";
+import { getOptions } from "./options.js";
+
+async function discoverMainFileUrl(outputDir: string): Promise<string> {
+  const tspUrl = await tryReadTspLocation(outputDir);
+  if (!tspUrl) {
+    throw new Error("No tsp-location.yaml found, and no main file specified");
+  }
+  Logger.debug(`Using base url from  tsp-location.yaml: ${tspUrl}`);
+  for (const mainFile of ["client.tsp", "main.tsp"]) {
+    let url = new URL(mainFile, tspUrl).toString();
+    Logger.debug(`Checking for file: ${url}`);
+    if (await doesFileExist(url)) {
+      Logger.debug(`Found main file: ${url}`);
+      return url;
+    }
+  }
+  throw new Error(`No main.tsp or client.tsp found for base url ${tspUrl}`);
+}
+
+async function downloadAndCompile({
+  rootUrl,
+  emitter,
+  noCleanup,
+  emitterOutputPath,
+}: {
+  rootUrl: string;
+  emitter: string;
+  noCleanup: boolean;
+  emitterOutputPath: string;
+}) {
+  const { moduleImports, fileTree } = await downloadTsp(rootUrl);
+  const { mainFilePath, files } = await fileTree.createTree();
+  const tempRoot = await createTempDirectory();
+  const srcDir = path.join(tempRoot, "src");
+  await writeFileTree(srcDir, files);
+
+  const emitterPackage = getEmitterPackage(emitter);
```

Comment:
Can you help me understand where to load the emitter version?

Reply:
Currently I don't try to figure out the emitter version and just grab latest. I was thinking that if we notice the output path has a folder of the format `azure-sdk-for-{something}` we could look for the `package.json` and get the emitter version from there. We could also choose the appropriate language emitter this way.

Reply:
My idea is to look for the emitter-package.json to get the emitter version. From my discussions with @lmazuel I think one of the goals here is to use this tool from the language repo clone, with something as simple as: `tsp-client generate`. If folks pass in certain flags or directories then we could customize it to the given parameters.

Reply:
FYI I'm working on this already and will open a PR against this branch once it's ready.

Reply:
@xirzec or @catalinaperalta, autorest.csharp was planning on floating the typespec dependencies and moving them to peerDependencies in the emitters, then .   This would allow typespec version flexibility for emitter consumers, but still pinning the version in azure-sdk-for-net in emitter-package.json as a devDependency:

[/eng/emitter-package.json](https://github.com/azure-sdk/azure-sdk-for-net/blob/auto-update-typespec-emitter-alpha-20230913.1/eng/emitter-package.json)
``` json
  "dependencies": {
    "@azure-tools/typespec-csharp": "0.2.0-alpha.20230913.1.x"
  },
  "devDependencies": {
    "@typespec/compiler": "0.47.1",
    "@typespec/versioning": "0.47.0",
    "@typespec/rest": "0.47.0",
    "@typespec/eslint-config-typespec": "0.47.0",
    "@azure-tools/typespec-client-generator-core": "0.33.0",
    "@typespec/http": "0.47.0",
    "@azure-tools/typespec-azure-core": "0.33.1"
  }
```

We also pin the versions in the accompanying [emitter-package-lock.json](https://github.com/azure-sdk/azure-sdk-for-net/blob/auto-update-typespec-emitter-alpha-20230913.1/eng/emitter-package-lock.json).

Will tsp-client use the lock file and devDependencies if they exists?

Reply:
I agree, this new tool will need to include the changes @hallipr is working on.

Reply:
We're currently using the `emitter-package.json` as is to install dependencies, so that should just work for you. Here's a tracking issue for supporting the lock file: https://github.com/Azure/azure-sdk-tools/issues/7057